### ISSUE TITLE
[REVIEW]: Coral: a parallel spectral solver for fluid dynamics and partial differential equations

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2978]
**Submitting author:** @benmql (<a href="http://orcid.org/0000-0001-6283-0382">BENJAMIN MIQUEL</a>)
**Repository:** <a href="https://github.com/BenMql/coral" target ="_blank">https://github.com/BenMql/coral</a>
**Version:** v1.1.12
**Editor:** @eloisabentivegna
**Reviewer:** @robertsawko, @rhaas80
**Archive:** <a href="https://doi.org/10.5281/zenodo.5458888" target="_blank">10.5281/zenodo.5458888</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/9cb3168716912f13021339e58117fdee/status.svg)](https://joss.theoj.org/papers/9cb3168716912f13021339e58117fdee)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/9cb3168716912f13021339e58117fdee"><img src="https://joss.theoj.org/papers/9cb3168716912f13021339e58117fdee/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/9cb3168716912f13021339e58117fdee/status.svg)](https://joss.theoj.org/papers/9cb3168716912f13021339e58117fdee)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/BenMql/coral" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@robertsawko & @rhaas80, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @eloisabentivegna know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/BenMql/coral">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@benmql) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/BenMql/coral">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@benmql) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @robertsawko, @rhaas80 it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1017/jfm.2020.485 is OK

MISSING DOIs

- 10.1103/physrevfluids.4.121501 may be a valid DOI for title: Convection driven by internal heat sources and sinks: Heat transport beyond the mixing-length or “ultimate” scaling regime

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
:wave: @robertsawko, please update us on how your review is going.

**Comment by whedon** [comment]
:wave: @rhaas80, please update us on how your review is going.

**Comment by eloisabentivegna** [comment]
@robertsawko, @rhaas80, could you post a review update?

**Comment by rhaas80** [comment]
comments working through installation:
* in the git clone command `git clone https://github.com/BenMql/coral.git .` I would either leave out the dot `.` or replace by `$PWD` which is easier to see
* `export CORAL_ROOT=$(PWD)` is incorrect and must be `$PWD` or `${PWD}` or `$(pwd)` since there is no `PWD` executable only, a `PWD` variable and a `pwd` executable.
* no link is provided for downloading MKL, would benice to have though Intel seems to restructure their website periodically making this harder
* MKL is not open-source (it is free-as-in-beer but not free-as-in-speech), would be nice to have alternative, also for non-Intel CPUs
* Decomp2d seems to have its own set of dependencies eg FFTE that the `Decomp2d_installation` file does not mention
* Decomp2d's makefile refers to `$(MKL_PATH)/lib/ia32` though by now `intel64` ought to be the default choice
* unclear if Decomp2d must be built against FFTW3 (coral is) or if using MKL's FFTW wrapper are ok
* FAILURE: at least the `clean` target assumes that `CORAL_ROOT` ends in a slash, which is not the case for `export CORAL_ROOT=$PWD`
* FAILURE: gfortran-10 fails with:
```
/src/sparse_tools/sparse_manipulations.f90:124:19:

  124 |                    ptr2B, A%ncol, dbeta, ptr2C, A%nrow)
      |                   1
......
  189 |                    ptr2B, A%ncol, dbeta, ptr2C, A%nrow)
      |                   2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (COMPLEX(8)/REAL(8)).
```
* needs either fix in code or `-fallow-argument-mismatch` in `MPIFLAGS` (proper code being preferred)
* OPTIONAL: parallel build using `make -j` fails since dependencies for module files are not recorded in `makefile`

**Comment by BenMql** [comment]
Thank you for your comments.  
+ `.`, `$PWD` and missing slashes issues have been fixed. 
+ the `2decomp` doc now states explicitly that a vanilla fftw is required.
+ A link to MKL has been added.
+ I share your enthusiasm for "BLAS/LAPACK-vendor" agnostic code. Down the road, my hope is that Coral can be linked to any BLAS/LAPACK library. For the time being though, I stick to `MKL` as some of their (non standard) routines are used.
+ working on the failures. This is interesting, I never detected them despite an heavy use of `pedantic` flags and standard-compliance checks (not with gfortran-10 though).

**Comment by rhaas80** [comment]
The argument mismatch is new by gfortran. Apparently such usage was never really allowed by the Fortran standard, but also never really enforced by compilers (though you may want to give eg the Cray compiler a try which in my experience is pickier than gfortran or ifort). You are certainly not the only code tripped up by this, eg https://www.google.com/search?q=gfortran-10+allow-mismatch&ie=utf-8&oe=utf-8 . For gfortran's release announcement about the issue, please see: https://gcc.gnu.org/gcc-10/changes.html#fortran

**Comment by BenMql** [comment]
Thank you for these details. This feature definitely helped catching a bug which fortunately had no consequences. (Some pointers used to be twice as long as intended, but fortunately their second half was never accessed.)
Anyway, this glitch is fixed now. Let me know if you run into more trouble while compiling.

**Comment by BenMql** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: :point_left:

**Comment by BenMql** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: :point_left:

**Comment by rhaas80** [comment]
In "2_First_run":
* typo" "After edition" should be "After editing"
* should indicate how much memory the sample run takes
* trying the test data on any number of MPI ranks > 1 I get a SEGFAULT with a backtrace and some information that seems to indicate that `opt_decomp` passed to `transpose_y_to_z` is incorect. Here's the information:
```
Thread 1 "coral_LP.exe" received signal SIGSEGV, Segmentation fault.
__memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:262
262     ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S: No such file or directory.
(gdb) up
#1  0x00007fffecd982fc in ?? () from /usr/lib/x86_64-linux-gnu/libopen-pal.so.40
(gdb) up
#2  0x00007fffecf268aa in ompi_datatype_sndrcv () from /usr/lib/x86_64-linux-gnu/libmpi.so.40
(gdb)
#3  0x00007fffecf732e3 in ompi_coll_base_alltoallv_intra_pairwise ()
   from /usr/lib/x86_64-linux-gnu/libmpi.so.40
(gdb)
#4  0x00007fffe65b0fa2 in ompi_coll_tuned_alltoallv_intra_dec_fixed ()
   from /usr/lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_coll_tuned.so
(gdb)
#5  0x00007fffecf2936b in PMPI_Alltoallv () from /usr/lib/x86_64-linux-gnu/libmpi.so.40
(gdb)
#6  0x00007fffed034d0b in pmpi_alltoallv__ () from /usr/lib/x86_64-linux-gnu/libmpi_mpifh.so.40
(gdb)
#7  0x00005555555de192 in decomp_2d::transpose_y_to_z_real (
    src=<error reading variable: value requires 2654208 bytes, which is more than max-value-size>,
    dst=<error reading variable: value requires 2654208 bytes, which is more than max-value-size>,
    opt_decomp=...) at transpose_y_to_z.f90:87
87               real_type, DECOMP_2D_COMM_ROW, ierror)
(gdb) print opt_decomp
$1 = ( xst = (1, 1, 49), xen = (72, 96, 96), xsz = (72, 96, 48), yst = (1, 1, 49), yen = (72, 96, 96), ysz = (72, 96, 48), zst = (1, 49, 1), zen = (72, 96, 96), zsz = (72, 48, 96), x1dist = (72), y1dist = (96), y2dist = (48, 48), z2dist = (48, 48), x1cnts = (331776), y1cnts = (331776), y2cnts = (165888, 165888), z2cnts = (165888, 165888), x1disp = (0), y1disp = (0), y2disp = (0, 165888), z2disp = (0, -322823200), x1count = 331776, y1count = 331776, y2count = 165888, z2count = 165888, even = .TRUE. )
```
where `opt_decomp%z2disp(2)` being `-322823200` seems like the likely culprit to me.

**Comment by rhaas80** [comment]
This seems to be due to assuming that `allocate` would zero initialize the allocated memory. I can make the SEGFAULT go away by adding
```
    decomp%x1cnts = 0
    decomp%y1cnts = 0
    decomp%y2cnts = 0
    decomp%z2cnts = 0
    decomp%x1disp = 0
    decomp%y1disp = 0
    decomp%y2disp = 0
    decomp%z2disp = 0
```
just before the `prepare_buffer` call.

**Comment by BenMql** [comment]
Roland, 

I have fixed the wiki `2_first_run`.

I am most grateful for your checks, comments and suggestions.
In `2decomp&fft`, the arrays of integers are now initialized to 0 as you suggested (available as the most recent commit to `https://github.com/benmql/2decomp_emptyFourierFourier'). I will run the code through valgrind as you suggest, probably some time next week.

**Comment by eloisabentivegna** [comment]
Thanks @rhaas80 for your thorough comments, and @BenMql for addressing them promptly!

@robertsawko, do you have any update to report at this stage?

**Comment by rhaas80** [comment]
> Thanks @rhaas80 for your thorough comments, and @BenMql for addressing them promptly!
@eloisabentivegna  Sorry about the long delay. I am unfortunately not yet done with the review.

**Comment by eloisabentivegna** [comment]
> > @eloisabentivegna  Sorry about the long delay. I am unfortunately not yet done with the review.

Thanks for your progress so far, @rhaas80. I can see your checklist is not complete, but your help up to this point is very appreciated.

**Comment by robertsawko** [comment]
Apologies for a radio silence. I am going to get back to reviewing this week. I have been reporting on GH issues of the software repository as I think that's the recommendation with JOSS. Made a few comments about the documentation. Unfortunately I run into an issue with the build on Arch Linux, but I think I should be able to resolve it now.

**Comment by robertsawko** [comment]
Hello @eloisabentivegna,

I have closed off my remaining issues with the documentation and installations - it all works well now. I went through the checklist and modified accordingly. Later this week I will check the visualisation to confirm the functionality and performance, but all looks good so far.

**Comment by BenMql** [comment]
Thank you all for your time and your valuable suggestions, which have spurred improvements in both the code, the doc, and the paper.
If you have more remarks, I am looking forward to hearing them!

**Comment by BenMql** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: :point_left:

**Comment by eloisabentivegna** [comment]
Thanks, @robertsawko! You did the right thing: specific change requests go in the package repository (you can of course create mentions to this issue in that venue, if it makes sense).

**Comment by eloisabentivegna** [comment]
@rhaas80, @robertsawko, could you update us on the status of your reviews, and state whether you're ready to make a final recommendation?

**Comment by BenMql** [comment]
Thank you both for your suggestions and corrections. I did not have time to implement them yet, and the sanitary measures in France are such that in all likelihood I will be able to resume my work on this project by the end of the month only. Apologies for the delay and thank you for your patience.

**Comment by eloisabentivegna** [comment]
@BenMql, thanks for the update. I will wait for your changes before pinging the reviewers again.

**Comment by eloisabentivegna** [comment]
@BenMql, were you able to address the reviewers' requests yet? Could you send an update?

**Comment by BenMql** [comment]
@eloisabentivegna , apologies for the long delay, thank you for your patience. We had a string of Covid19-related issues in the household (nationwide school closures, sickness, local school closure) that have proved rather incapacitating. Those all came to a resolution now.
I have started to address the issues raised by the reviewers and will send an update when finished---this should be relatively quick, say a couple days.

**Comment by eloisabentivegna** [comment]
Thanks for the update, @BenMql. I am sorry to hear about all the disruption.

Keep us posted!

**Comment by eloisabentivegna** [comment]
@BenMql, could you post an update?

**Comment by BenMql** [comment]
Dear all, 

The issue that have been thoroughly reported by @rhaas80  and @robertsawko  are now fixed. 

IMHO, only issue #13 still deserves some additional work: I tried my best to have a pedagogical and gradual approach for teaching the user how to enter PDEs in the solver Coral. I am still working to streamline the process. On the bright side, the user is now provided with 3 full benchmarks taken from some (recent) published references, which provide some starting points for implementing their own equations (with the help of the doc). 
 
On a different note, while testing on different machines since my last update, I triggered a bug of the DFT library with some recent compilers. After a while, it became clear that the culprits were optimization flags, which should be rolled back a little with recent gcc compilers when compiling `2decomp&fft`. This is now properly documented in the wiki. Perhaps more interestingly, this has prompted me to implement automated tests in Coral, that allows the user to test their compilation of the DFT library. @eloisabentivegna Apologies for the additional delay, this buggy behaviour was completely unexpected and was completely under the radar until very recently.

**Comment by eloisabentivegna** [comment]
Thanks for the update, @BenMql! I am glad this process has prompted you to further develop Coral (for one reason or another!).

@rhaas80, @robertsawko, do you have any comments?

**Comment by eloisabentivegna** [comment]
@rhaas80, @robertsawko, could you respond to @BenMql's latest changes? Do they address your requests?

**Comment by danielskatz** [comment]
Hi all - just to let you know, because @eloisabentivegna is taking leave, I'll be stepping in as editor on this submission - please let me know when there are issues.

**Comment by danielskatz** [comment]
👋 @rhaas80, @robertsawko - can you update me on your status?  Are your reviews currently blocked on anything from @BenMql?

**Comment by rhaas80** [comment]
@danielskatz @BenMql sorry for being so late, not blocked by @BenMql  just swamped with short term tasks right now.

**Comment by robertsawko** [comment]
@danielskatz, @BenMql and @eloisabentivegna. Sorry, I have also been swamped with short term tasks and I will be away next week for 1w. I will make sure to schedule some time to finish off this review on the week of the 19th of July. You will hear from me then. Really sorry this taking so much time.

**Comment by danielskatz** [comment]
@whedon remind @robertsawko in 1 week

**Comment by whedon** [comment]
Reminder set for @robertsawko in 1 week

**Comment by danielskatz** [comment]
@whedon remind @danielskatz in 1 week

**Comment by whedon** [comment]
Reminder set for @danielskatz in 1 week

**Comment by robertsawko** [comment]
Hello,

Apologies again for delay. I have now finished reviewing the code. Few minor points to consider below. Overall I think the code is an excellent contribution.

 * As far as I can see there are no automated tests for the code. I would suggest @BenMql considers whether any low-level tests can be proposed for some aspects of the solver and try some standard testing suites. If that is not possible, then I would suggest putting example outputs of the visualised quantities in [here](https://github.com/BenMql/coral/wiki/4_data_visualisation).
 *  I could not verify the performance claims as I do not have access to large enough machine currently, but @BenMql provides a reference to another publication where they document a performance study. I find this entirely sufficient and therefore I am considering this task completed.
 *  The table of content of the wiki pages is becoming a little long. I would consider simplifying it with a custom sidebar as described in [here](https://docs.github.com/en/communities/documenting-your-project-with-wikis/creating-a-footer-or-sidebar-for-your-wiki#creating-a-sidebar).

I believe several outstanding issues can be also closed now. For instance [this](https://github.com/BenMql/coral/issues/11) is working for me on the latest commit.

**Comment by whedon** [comment]
:wave: @robertsawko, please update us on how your review is going (this is an automated reminder).

**Comment by danielskatz** [comment]
@BenMql - it looks like you now have some things to look at, as @robertsawko has just three items left to check off in the review, which I think match the comments above

**Comment by danielskatz** [comment]
@rhaas80 - when do you think you will be able to make progress on the remaining items in your review?

**Comment by rhaas80** [comment]
@danielskatz @BenMql I am hosting a summer school this week, so cannot guarantee much progress before the end of the week, sorry.

**Comment by danielskatz** [comment]
👋 Hi all - Can I get an update on the status of this review? 

@rhaas80  - it looks like you opened some issues 2 weeks ago. Once these are resolved, will your review be complete?  Or are there other things you are also doing?

@robertsawko - What is needed for you to complete your review?  Are the items in https://github.com/openjournals/joss-reviews/issues/2978#issuecomment-886235689 from about a month ago still outstanding?

@BenMql - What is your progress on the issues raised by @rhaas80 and those in the comment from @robertsawko?

Thanks all!!

**Comment by rhaas80** [comment]
I just finished reading through the source. Looks good to me. Overall quite readable. For some files (most of the LP Coral file) there are haders:
```
 !============================================================================= 
 !                            C O R A L 
 !============================================================================= 
 ! 
 ! MODULE: LP_IMEX_timestepping
 !  
 !> @author 
 !> Ben Miquel, www.normalesup.org/~benmiquel, benjamin.miquel@tutanota.com 
 ! 
 ! DESCRIPTION 
 !> core routines and data structure implementing the time-marching schemes
 !> of Ascher-Ruuth-Spiteri 1997 (Applied Numerical Mathematics 25, pp151-167)
 !> [denoted below as ARS97].
 ! 
 !============================================================================= 
```
but not quite for all. Since these are very useful to get a rough idea what may be expected in a file, having them for all files would be great (but purely optional).

**Comment by rhaas80** [comment]
so once the two issues related to community contribution and quantitative tests are addressed I have no further objections.

**Comment by BenMql** [comment]
Thank you @rhaas80 and @robertsawko for the careful reviews. @danielskatz, I will be back in the office on Monday and will address the last suggestions promptly.

**Comment by BenMql** [comment]
:wave: all, 

I believe I have addressed all the remaining issues, which I summarize here:

+ both @rhaas80 and @robertsawko mentioned the lack of quantitative tests (e.g. [here](https://github.com/BenMql/coral/issues/20)). I have now gathered all the tests in a dedicated wiki page [here](./validation). For each benchmark, users are provided with a python routine so that the computation of global quantities is straightforward, and these values can readily be compared to reference values given in the same [wiki page](./validation).
+ both @rhaas80 and @robertsawko mentioned the lack of community guidelines (e.g. [here](https://github.com/BenMql/coral/issues/21). A `CONTRIBUTING.md` file has been added to fill this void and address this issue.
+ The third item missing in @robertsawko 's checklist is:  "Performance: If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)" This is a bit of a grey area, but perhaps this item could be checked off as there is indeed no claim for Coral itself (although `2decomp&fft` has been benchmarked by their authors). In the long term, a systematic performance study will be available on the wiki. For now, I do not have enough data points though.
+ @robertsawko mentioned the relevance of a custom sidebar (as the number of pages increases), which I have happily implemented. 
+ @rhaas80 mentioned the lack of consistency regarding headers in the source file, as an optional comment. This is a good point. More generally, these headers are part of the documentation of the code written in Doxygen. This documentation is still work in progress but will hopefully cover all source files and subroutines (at which point the headers will be homogeneous).

Apologies for this long interruption during Summer. Again, I want to express my gratitude for your suggestions which are very helpful and result in improved sources/doc. 
Let me know if I have missed something.

**Comment by danielskatz** [comment]
@robertsawko & @rhaas80 - please take a look at @BenMql's changes, and see if you can check off any remaining items. If not, please state what additional work you think is needed

**Comment by rhaas80** [comment]
contribution.md looks very good. While no automated tests are provided the manual tests now come with a textual description of the expected numerical output that lets users manually verify that their copy of CORAL works as expected.

I have no further comments that would delay publication.

**Comment by danielskatz** [comment]
👋 @robertsawko - How do things now look from your side?

**Comment by robertsawko** [comment]
Hello and thanks to @danielskatz for a ping on another email.

@BenMql thanks for addressing all of the comments. I think the wiki looks very clear now and examples are well described. I don't have any further comments. I've checked the remaining boxes.

**Comment by danielskatz** [comment]
Thanks @robertsawko

**Comment by danielskatz** [comment]
@BenMql - I'll proofread the paper shortly, then give you a few final steps (perhaps including suggested text changes) to do.

**Comment by danielskatz** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1017/jfm.2020.485 is OK
- 10.1103/physrevfluids.4.121501 is OK
- 10.1017/S0022112096002789 is OK
- 10.1103/PhysRevFluids.5.113702 is OK
- 10.1103/PhysRevE.70.056312 is OK
- 10.1103/PhysRevResearch.2.023068 is OK
- 10.1016/j.cpc.2019.107110 is OK
- 10.21105/joss.01071 is OK
- 10.1515/jnum-2012-0013 is OK
- 10.5334/jors.237 is OK
- 10.5334/jors.239 is OK

MISSING DOIs

- 10.1063/1.3058072 may be a valid DOI for title: Hydrodynamic and Hydromagnetic Stability

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@BenMql  - I've suggested some changes in https://github.com/BenMql/coral/pull/22 - please either merge these or let me know what you disagree with

**Comment by BenMql** [comment]
@danielskatz -- thank you for your edits, much appreciated. They have been merged.

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@BenMql - I missed one issue, which will be fixed if you merge https://github.com/BenMql/coral/pull/23 (I hope) - Please merge this, and run `@whedon generate pdf` to check if it looks ok.  If so, then please go ahead and
- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by BenMql** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@BenMql - you might want to change the bib entry for your J Fluid Mech paper to have your first initial rather than first name for consistency, but it's not a big deal either way

**Comment by BenMql** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02978/joss.02978/10.21105.joss.02978.pdf) :page_facing_up: :point_left:

**Comment by BenMql** [comment]
@danielskatz That looks better. The version tag is v1.1.12
What do people customarily do: register the upload on Zenodo as software or publication? (Apologies for the stupid question, but after all there is a paper and a code in the archive.)

**Comment by BenMql** [comment]
@danielskatz -- the archived Coral v1.1.12 has been uploaded on Zenodo and has received the DOI `10.5281/zenodo.5458888`

**Comment by danielskatz** [comment]
@whedon set v1.1.12 as version

**Comment by whedon** [comment]
OK. v1.1.12 is the version.

**Comment by danielskatz** [comment]
@whedon set 10.5281/zenodo.5458888 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5458888" target="_blank">10.5281/zenodo.5458888</a> is the archive.

**Comment by danielskatz** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1017/jfm.2020.485 is OK
- 10.1103/physrevfluids.4.121501 is OK
- 10.1017/S0022112096002789 is OK
- 10.1103/PhysRevFluids.5.113702 is OK
- 10.1103/PhysRevE.70.056312 is OK
- 10.1103/PhysRevResearch.2.023068 is OK
- 10.1016/j.cpc.2019.107110 is OK
- 10.21105/joss.01071 is OK
- 10.1515/jnum-2012-0013 is OK
- 10.5334/jors.237 is OK
- 10.5334/jors.239 is OK

MISSING DOIs

- 10.1063/1.3058072 may be a valid DOI for title: Hydrodynamic and Hydromagnetic Stability

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2569

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2569, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1434689883927457793) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2570
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.02978](https://doi.org/10.21105/joss.02978)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by danielskatz** [comment]
@BenMql - The DOI doesn't yet resolve for me - when it does, I will close this issue and we'll be done.

**Comment by BenMql** [comment]
Fantastic. Thank you for your assistance.
The DOI doesn't resolve yet for me neither, but it is probably only a question of time.

**Comment by danielskatz** [comment]
Congratulations to @BenMql (BENJAMIN MIQUEL)!!

(The DOI is now working, so we're done)

And thanks to @eloisabentivegna for editing, and @robertsawko and @rhaas80 for reviewing!
We couldn't do this without you!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02978/status.svg)](https://doi.org/10.21105/joss.02978)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02978">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02978/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02978/status.svg
   :target: https://doi.org/10.21105/joss.02978
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02978/status.svg)](https://doi.org/10.21105/joss.02978)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
